crates/ide/src/diagnostics/fixes.rs

```diff
@@ -22,18 +22,18 @@ use text_edit::TextEdit;
 
 use crate::{diagnostics::Fix, references::rename::rename_with_semantics, FilePosition};
 
-/// A [Diagnostic] that potentially has a fix available.
+/// A [Diagnostic] that potentially has some fixes available.
 ///
 /// [Diagnostic]: hir::diagnostics::Diagnostic
-pub(crate) trait DiagnosticWithFix: Diagnostic {
-    fn fix(&self, sema: &Semantics<RootDatabase>) -> Option<Fix>;
+pub(crate) trait DiagnosticWithFixes: Diagnostic {
+    fn fixes(&self, sema: &Semantics<RootDatabase>) -> Option<Vec<Fix>>;
```

Comment:
Is there an important invariant that the vector must be non-empty if this returns `Some`? If not I think this should just return `Vec<Fix>` without the option.

Reply:
As I explained in https://github.com/rust-analyzer/rust-analyzer/pull/8345#discussion_r607323890 as a second option, I think `Vec<Fix>` is better too. However functions inside `fix` currently make extensive use of `Try` trait and turning them into `Vec<Fix>` without `Option` doesn't seem easy to do.

e.g:
https://github.com/rust-analyzer/rust-analyzer/blob/10a243ea55565a0dd1de52f8f802c3e3a7bfef54/crates/ide/src/diagnostics/fixes.rs#L162-L176

Reply:
Alternatively, I think it also makes sense to change it to return something like `Option<HashMap<AssistId,Assist>>` which would possibly make it easier to test a specific Assist.